source/slang/slang-ast-base.h

```diff
@@ -140,25 +367,84 @@ class Val : public NodeBase
     // decide whether they need to do anything).
     Val* substituteImpl(ASTBuilder* astBuilder, SubstitutionSet subst, int* ioDiff);
 
-    bool equalsVal(Val* val);
+    bool equals(Val* val) const
+    {
+        return this == val || const_cast<Val*>(this)->resolve() == val->resolve();
+    }
 
     // Appends as text to the end of the builder
     void toText(StringBuilder& out);
     String toString();
 
     HashCode getHashCode();
-    bool operator == (const Val & v)
+    bool operator == (const Val & v) const
     {
-        return equalsVal(const_cast<Val*>(&v));
+        return equals(const_cast<Val*>(&v));
     }
 
     // Overrides should be public so base classes can access
     Val* _substituteImplOverride(ASTBuilder* astBuilder, SubstitutionSet subst, int* ioDiff);
-    bool _equalsValOverride(Val* val);
     void _toTextOverride(StringBuilder& out);
-    HashCode _getHashCodeOverride();
+
+    Val* _resolveImplOverride();
+
+    Val* resolveImpl();
+    Val* resolve();
+    ValNodeDesc getDesc();
+
+    Val* getOperand(Index index) const
+    {
+        SLANG_ASSERT(m_operands[index].kind == ValNodeOperandKind::ValNode);
+        return (Val*)m_operands[index].values.nodeOperand;
+    }
+
+    Decl* getDeclOperand(Index index) const
+    {
+        SLANG_ASSERT(m_operands[index].kind == ValNodeOperandKind::ASTNode);
+        return (Decl*)(m_operands[index].values.nodeOperand);
+    }
+
+    int64_t getIntConstOperand(Index index) const
+    {
+        SLANG_ASSERT(m_operands[index].kind == ValNodeOperandKind::ConstantValue);
+        return m_operands[index].values.intOperand;
+    }
+
+    Index getOperandCount() const { return m_operands.getCount(); }
+
+    template<typename ... TArgs>
+    void setOperands(TArgs... args)
+    {
+        m_operands.clear();
+        addOrAppendToNodeList(m_operands, args...);
+    }
+    template<typename ... TArgs>
+    void addOperands(TArgs... args)
+    {
+        addOrAppendToNodeList(m_operands, args...);
+    }
+    template<typename T>
+    void addOperands(OperandView<T> operands)
+    {
+        for (auto v : operands)
+            m_operands.add(ValNodeOperand(v));
+    }
+    List<ValNodeOperand> m_operands;
```

Comment:
This is great. You've made the whole `Val` hierarchy much more uniform. We could theoretically extend this to the entire AST with a bit (okay... a lot) more work.